base/int.jl

```diff
@@ -159,6 +159,7 @@ julia> signed(unsigned(-2))
 ```
 """
 unsigned(x) = convert(Unsigned, x)
+
 signed(x::Unsigned) = reinterpret(typeof(convert(Signed, zero(x))), x)
```

Comment:
Also, I wanted to ask: Does anyone know why `signed` takes `x::Unsigned`, but `unsigned` takes `x::BitSigned`? I'm _guessing_ it's to prevent `BigInt` from falling under the `reinterpret` implementation. I'm currently just copying that in the `::Type{T}` definitions -- does that seem reasonable?

Reply:
yes -- the only difference is the presence/absence of `BigInt`

Reply:
Now that you mention it, it is a little weird, because there's nothing about the docstring for `Unsigned` that specifies it needs to be `isprimitivetype`. Nothing stops me from defining

```julia
struct WeirdUnsigned <: Unsigned
    msg::String
    val::UInt16
end
WeirdUnsigned(x::Int) = WeirdUnsigned("missing", UInt16(x))
Base.Signed(x::WeirdUnsigned) = signed(x.val)

julia> signed(WeirdUnsigned(4))
ERROR: bitcast: expected primitive type value for second argument
Stacktrace:
 [1] reinterpret at ./essentials.jl:381 [inlined]
 [2] signed(::WeirdUnsigned) at ./int.jl:162
 [3] top-level scope at none:0
```

Of course this isn't a catastrophe: the error is sensible and one could always provide a more specialized definition. (If it returned a borked answer, that would be a genuine problem.) But of course the same could be said of `Signed`, and `reinterpret(UInt, BigInt(4))` throws the same error.

Reply:
Yeah, that's exactly the sort of thing I was referring to. Thanks for including an example.

> Of course this isn't a catastrophe: the error is sensible and one could always provide a more specialized definition. (If it returned a borked answer, that would be a genuine problem.) But of course the same could be said of `Signed`, and `reinterpret(UInt, BigInt(4))` throws the same error.

Totally, I agree. But I think that it's the inconsistency with `unsigned()` that's a little weird. I think _either_ we should change this to also accept only `BitUnsigned`, so that you'd get the correct behavior above, _OR_ we should loosen `unsigned()` to accept any kind of `Signed`, and just add an extra overload for `BigInt`, exactly as you'd have to do for `WeirdUnsigned` above. Does that seem reasonable to you?

Reply:
I've opened https://github.com/JuliaLang/julia/blob/master/base/int.jl#L170 to continue this conversation, since it's not contained to this PR! :)

Reply:
(oops, i pasted the wrong link before. I've opened https://github.com/JuliaLang/julia/pull/30445.)

Reply:
(ahhh omg sorry, wrong link again. fixed now. https://github.com/JuliaLang/julia/issues/30473)